[BB-548] Add basic structure for the metabase connector

mateovespConductor opened the pull request:
#### Description

- [ ] Bug fix
- [x] New feature

NOTE: The purpose of this connector is to use it as a library for metabase connectors in versions v0.49 and v0.56.
It does not have the databases resource that will be added in the versioned connectors; it only synchronizes users and groups.

#### Useful links:

- [Baton SDK coding guidelines](https://github.com/ConductorOne/baton-sdk/wiki/Coding-Guidelines)
- [New contributor guide](https://github.com/ConductorOne/baton/blob/main/CONTRIBUTING.md)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Adds a Metabase connector to sync users and groups, derive member/manager entitlements, produce user-to-group grants, support paging, and expose Metabase version and API rate-limit metadata.

- Configuration
  - Adds Metabase Base URL and API Key configuration with UI help and icon; API key marked secret.

- Tests
  - Adds unit tests for users, groups, entitlements, grants, pagination, version retrieval, and rate-limit scenarios.

- Chores
  - Promotes protobuf to a direct dependency.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mateovespConductor committed:
add basic structure

mateovespConductor commented on pkg/client/client.go:
```diff
@@ -0,0 +1,176 @@
+		withLimitParam(options.Limit),
+		withOffsetParam(options.Offset),
+	}
+	for _, opt := range opts {
+		opt(queryUrl)
+	}
+
+	_, rateLimitDesc, err := c.doRequest(ctx, http.MethodGet, queryUrl, &res, nil)
+	if err != nil {
+		return nil, "", rateLimitDesc, fmt.Errorf("failed to fetch users: %w", err)
+	}
+
+	nextToken := getNextPageToken(res.Offset, res.Limit, res.Total)
+
+	return res.Data, nextToken, rateLimitDesc, nil
+}
```
The Metabase API actually returns a paginated response wrapped in an object with a data field, as shown below:
{
  "data": [...],
  "total": 5,
  "limit": 1,
  "offset": 2
}
So the UsersQueryResponse struct with a Data field is required — it matches the real API payload and works correctly in the connector
Model:
// UsersQueryResponse models the paginated response for user listings in Metabase.
type UsersQueryResponse struct {
	Data   []*User `json:"data"`
	Total  int     `json:"total"`
	Limit  int     `json:"limit"`
	Offset int     `json:"offset"`
}

mateovespConductor commented on pkg/connector/users.go:
```diff
@@ -2,34 +2,134 @@ package connector
+// Placing this in groups would require iterating over all users for each group,
+// which is costly and unnecessary.
+func (u *userBuilder) Grants(ctx context.Context, resource *v2.Resource, _ *pagination.Token) ([]*v2.Grant, string, annotations.Annotations, error) {
+	ann := annotations.New()
+	allMemberships, rateLimitDesc, err := u.client.ListMemberships(ctx)
+	if rateLimitDesc != nil {
+		ann.WithRateLimiting(rateLimitDesc)
+	}
+	if err != nil {
+		return nil, "", ann, fmt.Errorf("failed to list memberships: %w", err)
+	}
+
+	userMemberships, ok := allMemberships[resource.Id.Resource]
+	if !ok {
+		return nil, "", ann, nil
+	}
+
+	grants := make([]*v2.Grant, 0, len(userMemberships))
+	for _, membership := range userMemberships {
+		groupResource := &v2.Resource{
+			Id: &v2.ResourceId{
+				ResourceType: groupResourceType.Id,
+				Resource:     strconv.Itoa(membership.GroupID),
+			},
+		}
+
+		role := memberPermission
+		if membership.IsGroupManager {
+			role = managerPermission
+		}
+
+		grants = append(grants, grant.NewGrant(
+			groupResource,
+			role,
+			resource.Id,
+		))
+	}
+
+	return grants, "", ann, nil
+}
```
ListMemberships() is a GET request and the Baton SDK’s HTTP client caches GET responses by default. So the memberships are already cached during the sync, and we don’t make O(n²) calls. 
https://github.com/ConductorOne/baton-sdk/blob/17596013a933501434d3fce3b9e7221cfbb19298/pkg/uhttp/wrapper.go#L344

mateovespConductor committed:
Improve documentation and add corrections from coderabbit

luisina-santos commented on .github/workflows/ci.yaml:
```diff
@@ -7,18 +7,18 @@ jobs:
         uses: actions/setup-go@v5
         with:
-          go-version: 1.24.x
+          go-version: 1.25.x
```
follow the pattern of this PR to avoid this go version updated in the future > https://github.com/ConductorOne/baton-template/pull/52

luisina-santos commented on .github/workflows/ci.yaml:
```diff
@@ -37,42 +37,35 @@ jobs:
-          baton-entitlement: 'entitlement:group:1234:member'
-          baton-principal: 'user:9876'
-          baton-principal-type: 'user'
+#  ** TODO Attempt to run both versions locally using curl and documentation endpoints to obtain an API key that allows the connector to synchronize and run the test automatically.
```
could we run tests using always latest version of metabase here?

luisina-santos commented on .github/workflows/main.yaml:
```diff
@@ -10,18 +10,18 @@ jobs:
         uses: actions/setup-go@v5
         with:
-          go-version: 1.24.x
+          go-version: 1.25.x
```
if you update ci workflow following changes on this PR you should be able to remove this file 

https://github.com/ConductorOne/baton-template/pull/52

luisina-santos commented on README.md:
```diff
@@ -7,7 +7,73 @@
 
 # Prerequisites
-No prerequisites were specified for `baton-metabase`
+This connector requires the open source version Metabase v0.49.x or v0.56.x to be downloaded in order to function correctly.
```
this connector can be used in any version post v0.49 since needs API KEY support, right? should not be strictly attached to those specific versions

luisina-santos commented on pkg/client/client.go:
```diff
@@ -0,0 +1,176 @@
+		}
+		bodyStr := strings.TrimSpace(string(bodyBytes))
+
+		var errResp ErrorResponse
```
I think this error response can be added as uhttp option as we are doing here: https://github.com/ConductorOne/baton-contentful/blob/9f8ec9dc8b4ebb8febceb6b2d2cc3d0f8151f4ac/pkg/client/teams.go#L27

mateovespConductor replied:
The Metabase API does not always return a JSON-formatted error body. Using the WithErrorResponse helper would fail in such cases:
https://github.com/ConductorOne/baton-sdk/blob/17596013a933501434d3fce3b9e7221cfbb19298/pkg/uhttp/wrapper.go#L219

So I implemented manual error parsing in doRequest: first it tries to parse the response as JSON, and if that fails, it falls back to using the HTTP status text. In this case, I did it this way because Metabase provides short, unformatted text errors (such as “Unauthorized”) which report correctly, without breaking the connector or logging large HTML responses.

luisina-santos commented on pkg/client/client.go:
```diff
@@ -0,0 +1,176 @@
+		withLimitParam(options.Limit),
+		withOffsetParam(options.Offset),
+	}
+	for _, opt := range opts {
```
ReqOpt should be added as parameter of doRequest so you avoid doing this for iteration each time you need to use pagination options

luisina-santos commented on pkg/config/config.go:
```diff
@@ -5,19 +5,38 @@ import (
+	MetabaseBaseUrl = field.StringField(
+		"metabase-base-url",
+		field.WithRequired(true),
+		field.WithIsSecret(false),
```
you can omit this line if you are setting default value

luisina-santos reviewed (approved):
Minor comments, please address before merge! Well done!

mateovespConductor committed:
Add suggestions:
- Improve client to accept pagination in doRequest.
- Update documentation
- Clean up CI: improve ci.yaml and remove obsolete main.yaml

FeliLucero1 commented on pkg/client/client.go:
```diff
@@ -0,0 +1,172 @@
+func (c *MetabaseClient) ListUsers(ctx context.Context, options PageOptions) ([]*User, string, *v2.RateLimitDescription, error) {
+	var res UsersQueryResponse
+
+	queryUrl, err := url.Parse(c.baseURL + getUsers)
```
I prefer using path.Join

mateovespConductor committed:
- Make resource types public.
- Add functionality in the client to obtain the initial metabase version and use it in versioned connectors to validate that the correct version is being used.

mateovespConductor committed:
Change group permissions constants to public

mateovespConductor committed:
Improve the client to use JoinPath
